### PR TITLE
Rewrite `writer-json` using Boost.JSON

### PR DESCRIPTION
**Pull request by lzaoral**
CMake changes expect a build of Boost with Boost.JSON.

### CODE REVIEW

File: src/lib/writer-json.cc [hunk 1]
```diff
@@ -40,7 +42,104 @@ static inline std::string sanitizeUTF8(const std::string &str)
...
+        os << "null";
+        break;
+    }
+
+    if (indent->empty())
+        os << "\n";
+}
+
+// TODO: This should not necessary!  TScanProps should be able to contain
+// any type so that no conversions here are needed.
```
Comment (kdudka) [inline src/lib/writer-json.cc:128]: I would not worry too much.  Most of the properties come from `scan.ini`, which is not a type-aware format anyway:
```
[scan]
tool = csmock
tool-version = csmock-3.3.5.20230203.170717.ga8a9617.ini_strip-1.fc36
tool-args = '/usr/bin/csmock' '-r' 'fedora-37-x86_64' '-f' '/home/kdudka/git/csmock/rpm-inst-trash/csmock-3.3.5.20230202.094355.g16818ee.dirty-1.fc36.src.rpm' '-t' 'unicontrol,gitleaks'
host = f36
store-results-to = /home/kdudka/git/csmock/csmock-3.3.5.20230202.094355.g16818ee.dirty-1.fc36.tar.xz
time-created = 2023-02-03 17:07:39
enabled-plugins = gitleaks, unicontrol
mock-config = fedora-37-x86_64
project-name = csmock-3.3.5.20230202.094355.g16818ee.dirty-1.fc36
known-false-positives = /usr/share/csmock/known-false-positives.js
analyzer-version-gitleaks = 8.15.1
analyzer-version-unicontrol = 0.0.2
time-finished = 2023-02-03 17:08:48
exit-code = 0
```

File: src/lib/writer-json.cc [hunk 2]
```diff
@@ -40,7 +42,104 @@ static inline std::string sanitizeUTF8(const std::string &str)
     return convert_string<char>(str.data(), str.data() + str.size());
 }
 
-typedef SharedStringPTree PTree;
+static void pretty_print(std::ostream &os, const value &jv, std::string* indent = nullptr)
```
Comment (kdudka) [inline src/lib/writer-json.cc:45]: Could you please put encoding of objects and array to separate functions to make the code easier to follow?

The recursive implementation could be a problem if an output format supported by csdiff allowed to control nesting level by the input data.  This is luckily not the case now.  At least I hope so...

File: src/lib/writer-json.cc [hunk 3]
```diff
@@ -40,7 +42,104 @@ static inline std::string sanitizeUTF8(const std::string &str)
     return convert_string<char>(str.data(), str.data() + str.size());
 }
 
-typedef SharedStringPTree PTree;
```
Comment (kdudka) [inline src/lib/writer-json.cc:43]: Thanks to this change, you can also drop the include of `shared-string-ptree.hh` and consequently remove the `shared-string-ptree.hh` header file and the whole `shared-string` module from the project.

**Review by kdudka — CHANGES_REQUESTED** [review]
I like this implementation.  It performs much better in terms of memory consumption than the current implementation.  A few suggestions inline...

**Comment by lzaoral** [comment]
Newer Boost in EPEL is tracker here: https://bugzilla.redhat.com/show_bug.cgi?id=2165584#

File: src/lib/writer-json.cc (hunk 2 above)
Comment (lzaoral) [reply src/lib/writer-json.cc]: Addressed in 944abf1.

File: src/lib/writer-json.cc (hunk 3 above)
Comment (lzaoral) [reply src/lib/writer-json.cc:43]: Addressed in 5706789.

File: src/lib/writer-json.cc (hunk 1 above)
Comment (lzaoral) [reply src/lib/writer-json.cc:145]: Thanks for the explanation!

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -24,15 +24,8 @@ set(cssort      "${CMAKE_BINARY_DIR}/src/cssort")
...
-# unify the JSON output generated by different version of Boost
-set(jsfilter "sed")
-set(jsfilter "${jsfilter} -e ':a;N;$!ba;s|:\\\\\\\\n *\\\\\\\\[|: [|g'")
-
 # eliminate csdiff's version string in the output
-set(jsfilter "${jsfilter} -e 's|\"version\": \"[^\"]*\"|\"version\": \"\"|g'")
-
-# yet another conversion needed for scan properties on EPEL-7
-set(jsfilter "${jsfilter} -e ';s|:\\\\\\\\n *{|: {|g'")
+set(jsfilter "sed -e 's|\"version\": \"[^\"]*\"|\"version\": \"\"|g'")
```
Comment (kdudka) [inline tests/CMakeLists.txt:28]: I guess we can remove these filters also from `tests/test-lib.sh`?

Comment (lzaoral) [reply tests/CMakeLists.txt:28]: Sure, addressed in 30964cd.

**Review by kdudka — APPROVED** [review]
Thanks for the update!  I think we can merge this...

**Comment by lzaoral** [comment]
Thank you, @kdudka! Merging.

**Pull request merged**